### PR TITLE
feat(runtime): proxy form associated custom element lifecycle callbacks

### PR DESCRIPTION
**Pull request by HansClaasen**
## What is the current behavior?
Currently a Stencil component can not make use of the following form associated custom element lifecycle callbacks:

- `formAssociatedCallback`
- `formResetCallback`
- `formDisabledCallback`
- `formStateRestoreCallback`

See more on these callbacks on https://web.dev/articles/more-capable-form-controls

GitHub Issue Number: N/A

## What is the new behavior?

With this patch we update the proxy class to propagate the function calls to the component, if the Stencil component has a `formAssociated` flag set.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

I enhanced an e2e test to set the default aria label for the component to a specific value which I then assert in the test. For simplicity reasons I refrained adding a test for every supported callback given they are all proxied the same way. Happy to add the tests though if desired.

## Other information

There might be things, e.g. missing types or compiler changes, that I might have overseen. Please point me to missing pieces if so.

Also I am not sure how much we want to optimize the compiler here. Currently code will added to the stencil core even though none of the components are using any of these form associated callbacks. I can add compiler flags for each one of them to ensure we actually ship a minimum necessary amount of code.

### CODE REVIEW

**Review by alicewriteswrongs — CHANGES_REQUESTED** [review]
👋 thanks for this contribution!

I just took it for a quick test-drive and everything looks good, I was able to modify my [form-associated example](https://github.com/alicewriteswrongs/stencil-element-internals-example) to use the `formAssociatedCallback` and it worked great!

I just noted a few things in the code

File: src/runtime/runtime-constants.ts [hunk 1]
```diff
@@ -61,3 +61,10 @@ export const HYDRATE_CHILD_ID = 'c-id';
 export const HYDRATED_CSS = '{visibility:hidden}.hydrated{visibility:inherit}';
 
 export const XLINK_NS = 'http://www.w3.org/1999/xlink';
+
+export const FORM_ASSOCIATED_CUSTOM_ELEMENT_CALLBACKS = [
+  'formAssociatedCallback',
+  'formResetCallback',
+  'formDisabledCallback',
+  'formStateRestoreCallback'
+]
```
[the remark this answers is not in the document]
Comment (HansClaasen) [reply src/runtime/runtime-constants.ts:70]: Good idea!

**Comment by HansClaasen** [comment]
@alicewriteswrongs thanks for the review. I addressed your comment.

**Comment by rwaskiewicz** [comment]
@HansClaasen It looks like the Karma tests are failing here for all browsers we test against:
```
Firefox 118.0 (Windows 10) form associated should trigger form associated custom element lifecycle callbacks FAILED
	Expected undefined to be 'formAssociated called'.
	<Jasmine>
	global.wrappers["/home/runner/work/stencil/stencil/test/karma/test-app/form-associated/karma.spec.ts"]</</</</<@test-app/form-associated/karma.spec.js:67:38
	step@test-app/form-associated/karma.spec.js:33:23
	verb/<@test-app/form-associated/karma.spec.js:14:53
	global.wrappers["/home/runner/work/stencil/stencil/test/karma/test-app/form-associated/karma.spec.ts"]</__awaiter</<@test-app/form-associated/karma.spec.js:8:71
	global.wrappers["/home/runner/work/stencil/stencil/test/karma/test-app/form-associated/karma.spec.ts"]</__awaiter<@test-app/form-associated/karma.spec.js:4:12
	global.wrappers["/home/runner/work/stencil/stencil/test/karma/test-app/form-associated/karma.spec.ts"]</</<@test-app/form-associated/karma.spec.js:63:98
	<Jasmine>
```

These tests historically can be a bit of a pain to debug.  To make it a bit easier, I suggest building the compiler (`npm run build`) and navigating to `test/karma`. From there, you can start a Stencil dev server with `npm start` and navigate to http://localhost:3333/test-www/form-associated (the port may be different, check the output of starting the dev server to make sure it matches). From there, you should be able to poke at the component and start debugging. 

The format check is also failing, but that should be an easy fix 🙂 `npm run prettier`

**Comment by HansClaasen** [comment]
@rwaskiewicz thanks for the tip 👌 I was able to debug the issue and resolve it. The problem was that the prototype tweak was only applied if the host element had members. I moved the change outside of that if statement and it started to work for the test component.

**Review by alicewriteswrongs — APPROVED** [review]
I think this looks good! There's some lint / formatting issues to resolve, but once that's fixed I think this is good to go

**Review by tanner-reits — APPROVED** [review]
Think this looks good! Only ask would be to resolve some of the new SNC violations this introduced, but I won't consider that a blocker

**Pull request merged**